### PR TITLE
wpa_supplicant: allow both imperative and declarative networks

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

:warning: __Warning:__ this PR contains a patch for `wpa-supplicant`, an IMHO pretty sensitive software, also this is an experiment from me to see how well both static and imperative networks work in `wpa_supplicant`. So before considering this mergable, a few people with more C experience than I have should take a look!

------

For a while now it's possible to specify an additional config file in
`wpa_supplicant`[1]. In contrast to the file specified via `-c` this was
supposed to be used for immutable settings and not e.g. additional
networks.

However I'm a little bit unhappy about the fact that one has to choose
between a fully imperative setup and a fully declarative one where the
one would have to write credentials for e.g. WPA2-enterprise networks
into the store.

The primary problem with the current state of `wpa_supplicant` is that
if the `SAVE_CONFIG` command is invoked (e.g. via `wpa_cli`), all known
networks will be written to `/etc/wpa_supplicant.conf` and thus all
declarative networks would get out of sync with the declarative
settings.

To work around this, I had to change the following things:

* The `networking.wireless`-module now uses `-I` for declarative config,
  so the user-controlled mode can be used along with the
  `networks`-option.

* I added an `ro`-field to the `ssid`-struct in the
  `wpa_supplicant`-sources. This will be set to `1` for each network
  specified in the config passed via `-I`.

  Whenever config is written to the disk, those networks will be
  skipped, so changes to declarative networks are only temporary.

[1] https://w1.fi/cgit/hostap/commit/wpa_supplicant?id=e6304cad47251e88d073553042f1ea7805a858d1

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113716` at 9ef9c044 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/da47aa7877dd6c4d62e551c196f8c5be">wifish</a></li></li>
    <li><a href="https://gist.github.com/4271c469d9c157c45c95143dd2a3fd95">wpa_supplicant</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/os-specific/linux/wpa_supplicant/default.nix:17:5:
```
   |
17 |     (fetchurl {
   |     ^
```
</li>
  </ul>
</details>

**Comment by bb2020** [comment]
Is using `wpa_supplicant` with an external password file a solution to "passwords-in-store" problem?
```
networking.wireless.extraConfig = "ext_password_backend=file:/path/to/passwords.conf";
```

**Comment by Ma27** [comment]
Interesting, I wasn't aware of this option, I'd guess so :)
However there's another usecase I'd like to cover: sometimes I only need a wifi net temporarily and then I really don't want to nixos-rebuild my system. Also, it can be helpful to initially configure a net with e.g. wpa_gui and "persist" it later on :)

**Comment by bb2020** [comment]
It is against declarative nature of the service if config is read from an external file. It may crash if the external file is incorrectly formatted etc. I think this should be solved at `wpa_gui` side not at `wpa_supplicant` side. However I only used wired network, so I am not proficient in wireless configuration.

**Comment by Ma27** [comment]
> I think this should be solved at wpa_gui side not at wpa_supplicant side

The thing is, both `wpa_cli` and `wpa_gui` send a `SAVE_CONFIG` to `wpa_supplicant` with a payload, so the actual logic regarding how changes are persisted is implement in `wpa_supplicant`. Or am I misunderstanding the approach you're thinking of? :) 

> It is against declarative nature of the service if config is read from an external file

Yes. OTOH I think that a boundary of declarativity is reached: while I like to configure frequently used networks declaratively, I don't want to rebuild my system for let's say a hotel wifi.

Would you prefer a solution where this behavior is opt-in?

> It may crash if the external file is incorrectly formatted et

That's a valid concern!

Please note that I don't expect this to be merged kinda soon. This was mostly intended as an "experiment" for now. Before this hits upstream master, a few more people should review this (especially the C patch). To give this some kind of basic test coverage, I'll use this on my laptop for now to see if I encounter any issues.

**Comment by SuperSandro2000** [comment]
I marked the PR as draft to better reflect the experimental status. Feel free to undraft at any time.

**Comment by bb2020** [comment]
> Please note that I don't expect this to be merged kinda soon. This was mostly intended as an "experiment" for now. Before this hits upstream master, a few more people should review this (especially the C patch). To give this some kind of basic test coverage, I'll use this on my laptop for now to see if I encounter any issues.

I think patching `wpa_supplicant` itself is kinda extreme for this problem. Maybe you can try overriding `systemd` service's `script` attribute to combine `configFile` and `/etc/wpa_supplicant.conf` into a single file and then feed that file to `-c` argument.

**Comment by Ma27** [comment]
> Maybe you can try overriding systemd service's script attribute to combine configFile and /etc/wpa_supplicant.conf into a single file and then feed that file to -c argument.

Can you elaborate? As soon as wpa_supplicant writes to this file, declarative changes will be modified.

**Comment by bb2020** [comment]
There is also `networking.supplicant` option. Is it possible to declare declarative networks in `extraConf` and set `configFile.path` to include temporary networks? This requires `networking.wireless` to be disabled, I presume.

**Comment by Ma27** [comment]
For some reason, that one is duplicated and should be removed actually: https://github.com/NixOS/nixpkgs/issues/22401

**Comment by bb2020** [comment]
Well, it is still available in `21.09` so you can try it.

**Comment by Ma27** [comment]
> Is it possible to declare declarative networks in extraConf and set configFile.path to include temporary networks?

AFAIU no. This path will be passed to `wpa_supplicant` with `-I` and then we have the exact same issue as described above.

**Comment by bb2020** [comment]
It seems `configFile.path`  is passed with `-c` [here](https://github.com/NixOS/nixpkgs/blob/d4189f68fdbe0b14b7637447cb0efde2711f4abf/nixos/modules/services/networking/supplicant.nix#L29). `extraConf` is written to a file and passed with `-I`.

**Comment by Ma27** [comment]
Yes. And temporary networks have to be written to writable file. Hence `configFile.path` passed via `-c`. `wpa_supplicant` doesn't write to the file passed via `-I`.

**Comment by bb2020** [comment]
I think this should work like this `supplicant."wlan0".configFile.path = "/etc/wpa_suppliant.conf"`. And then you can set declarative configurations in `supplicant."wlan0".extraConf`.

**Comment by Ma27** [comment]
This is the config passed via `-I` in the end. When you modify this network now in e.g. a CLI or GUI interface, it will be *copied* to `/etc/wpa_supplicant`. Since the list of known SSIDs is a linked list internally, the modified value will take precedence over the one from your NixOS config. And due to that property, the idea of using `-I` for declarative config rather than `-c` isn't a good one.

**Comment by bb2020** [comment]
Manual for `networking.supplicant.<name>.extraConf` says `Options defined here have precedence over options in configFile`.

**Comment by Ma27** [comment]
I don't know where this claim comes from, but I actually confirmed that overriding config in `wpa_gui` and having it saved to `/etc/wpa_supplicant.conf` takes precedence. I mean I hoped this wouldn't be the case as this would've saved the time I invested in patching wpa_supplicant :D

**Comment by bb2020** [comment]
I think this is even better. If wifi password is changed temporarily, it can be overridden in the `conf` file without rebuilding the system. Especially at different places with the same SSID. Don't you think?

**Comment by Ma27** [comment]
I think that silently overriding declarative (as in "single source of truth") values is not really desirable.

**Comment by bb2020** [comment]
It sounded dangerous when you say it like that.

**Comment by bb2020** [comment]
`wpa_config_write` re-saves network entries coming from the constant file (passed with `-I`) into the generic file (passed with `-c`), right? Actually this is an upstream issue. Maybe we should report this to `wpa_supplicant` developers?

**Comment by Ma27** [comment]
IIRC the file passed with `-I` isn't supposed to be used for networks (even though that's possible). I mainly decided to use this mechanism for "our needs" :)

File: pkgs/os-specific/linux/wpa_supplicant/0001-Implement-read-only-mode-for-ssids.patch
Comment (bb2020) [inline pkgs/os-specific/linux/wpa_supplicant/0001-Implement-read-only-mode-for-ssids.patch:59]: `if (ssid->key_mgmt == WPA_KEY_MGMT_WPS || ssid->temporary || ssid->ro)`

See how they used `temporary`.

**Comment by bb2020** [comment]
Maybe the best idea is, to introduce an option like `networking.wireless.allowImperativeNetworks` and making it disabled by default. Also introduce a new derivation, something like `wpa_supplicant-ro` (similar to `gui.nix`) and then replace `pkgs.wpa_supplicant` with `pkgs.wpa_supplicant-ro` if `allowImperativeNetworks` is `true`.

**Comment by Ma27** [comment]
That's what I meant with "opt-in" solution :) Sounds reasonable to me.

**Comment by bb2020** [comment]
I don't use wireless, but for those who do, there is definitely a problem. At first I didn't like the idea of patching `wpa_supplicant` itself, but then again failed to offer a more viable solution. So I thought about creating a new derivation, like `wpa_supplicant_ro` that includes C code patches. This way, an untouched version of `wpa_supplicant` would still be available in `nixpkgs`.

**Comment by Ma27** [comment]
Yeah I agree with that and I'll do this change before it gets merged. For now, I'd like to get a bit of feedback about the overall approach to know whether I should continue focusing on getting this into nixpkgs :)

**Comment by Ma27** [comment]
Ok, just implemented the new behavior as opt-in now :)

**Comment by bb2020** [comment]
This looks great! Have you seen [this](https://github.com/NixOS/nixpkgs/pull/113716#discussion_r598826542)?

What about naming the option as something like `allowImperativeNetworks` or `allowAuxiliaryImperativeNetworks`?
Because declarative networks are allowed regardless of the setting :laughing:

File: nixos/modules/services/networking/wpa_supplicant.nix
Comment (bb2020) [inline nixos/modules/services/networking/wpa_supplicant.nix:55]: Is `default = false` needed here?

Comment (Ma27) [reply nixos/modules/services/networking/wpa_supplicant.nix:55]: No :)

**Comment by Ma27** [comment]
> What about naming the option as something like allowImperativeNetworks or allowAuxiliaryImperativeNetworks?

Good idea, fixed in my last push :)

**Comment by bb2020** [comment]
Awesome. I think this PR is ready for prime time. It still can't beat ethernet though :smile:

**Comment by Ma27** [comment]
> I think this PR is ready for prime time.

Yeah. I'm running it for >2 months now without problems.

> It still can't beat ethernet though

That's true :) 

@bb2020 I'll mark this as "ready for review" now. Will wait for a at least a week now.

**Comment by bb2020** [comment]
I don't have wifi to test it, but the code seems fine to me. It is disabled by default and an untouched version of `wpa_supplicant` is still in `nixpkgs` so there is not much to complain.

I wish there was a solution without patching native code, but for now, this looks like the only option.

**Review by bb2020 — APPROVED** [review]

**Review by lheckemann — APPROVED** [review]

**Comment by Ma27** [comment]
I'm going to merge this now.

I discussed a bit about the patch with my colleague Linus last week and we agreed that even though I'm not very confident that this will be merged into upstream, I'll give it a try and file a patch: https://lists.infradead.org/pipermail/hostap/2021-May/039618.html

**Pull request merged**
